ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -585,6 +581,75 @@ public func first<T, E>(producer: SignalProducer<T, E>) -> Result<T, E>? {
 	return result
 }
 
+/// Starts the producer, then blocks, waiting for events: Next and Completed.
+/// When a single value or error is sent, the returned `Result` will represent
+/// those cases. However, when no values are sent, or when more than one value
+/// is sent, `nil` will be returned.
+public func single<T, E>(producer: SignalProducer<T, E>) -> Result<T, E>? {
+	let semaphore = dispatch_semaphore_create(0)
+	var result: Result<T, E>?
+
+	producer
+		|> take(2)
+		|> start(next: { value in
+			if result != nil {
+				// Move into failure state after recieving another value.
+				result = nil
+				return
+			}
+
+			result = success(value)
+		}, error: { error in
+			result = failure(error)
+			dispatch_semaphore_signal(semaphore)
+		}, completed: {
+			dispatch_semaphore_signal(semaphore)
+			return
+		})
+
+	dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER)
+	return result
+}
+
+/// Starts the producer, then blocks, waiting for the last value.
+public func last<T, E>(producer: SignalProducer<T, E>) -> Result<T, E>? {
+	let semaphore = dispatch_semaphore_create(0)
+	var result: Result<T, E>?
+
+	producer
+		|> takeLast(1)
+		|> start(next: { value in
+			result = success(value)
+			dispatch_semaphore_signal(semaphore)
+		}, error: { error in
+			result = failure(error)
+			dispatch_semaphore_signal(semaphore)
+		}, completed: {
+			dispatch_semaphore_signal(semaphore)
+			return
+		})
+
+	dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER)
+	return result
+}
+
+/// Starts the producer, then blocks, waiting for completion.
+public func wait<T, E>(producer: SignalProducer<T, E>) -> Result<(), E> {
+	let semaphore = dispatch_semaphore_create(0)
+	var result: Result<(), E>!
+
+	producer.start(error: { error in
```

Comment:
This could probably be implemented in terms of one of the others, if you're interested in trying it.